README.md
# Tracing helper library

The library creates a configured tracer instance.


## Usage

Create a tracer instance and set it as a global tracer:

```go
import (
		"github.com/opentracing/opentracing-go"
        "gerrit.o-ran-sc.org/ric-plt/tracelibgo/pkg/tracelibgo"
        ...
)

tracer, closer := tracelibgo.CreateTracer("my-service-name")
defer closer.Close()
opentracing.SetGlobalTracer(tracer)
```

Serialize span context to a byte array that can be sent
to another component via some messaging. For example, using
the RMR library. The serialization uses JSON format.

```go
	carrier := make(map[string]string)
	opentracing.GlobalTracer().Inject(
			span.Context(),
			opentracing.TextMap,
			opentracing.TextMapCarrier(carrier))
	b, err := json.Marshal(carrier) // b is a []byte and contains serilized span context
```

Extract a span context from byte array and create a new child span from it.
The serialized span context is got, for example, from the RMR library.

```go
	var carrier map[string]string
	err = json.Unmarshal(data, &carrier) // data is []byte containing serialized span context
	if err != nil {
		...
	}
	context, err := opentracing.GlobalTracer().Extract(opentracing.TextMap, opentracing.TextMapCarrier(carrier))
	if err != nil {
		...
	}
	span := opentracing.GlobalTracer().StartSpan("go test span", opentracing.ChildOf(context))
```

## Configuration

The trace library currently supports only [Jaeger](https://www.jaegertracing.io/) [golang client](https://github.com/jaegertracing/jaeger-client-go) tracer implementation.
The configuration is done using environment variables:

| environment variable         | values                              | default        |
| ---------------------------- |------------------------------------ | -------------- |
| TRACING_ENABLED              | 1, true, 0, false                   | false          |
| TRACING_JAEGER_SAMPLER_TYPE  | const, propabilistic, ratelimiting  | const          |
| TRACING_JAEGER_SAMPLER_PARAM | float                               | 0.001          |
| TRACING_JAEGER_AGENT_ADDR    | IP addr[:port]                      | 127.0.0.1:6831 |
| TRACING_JAEGER_LOG_LEVEL     | all, error, none                    | none           |

Meaning of the configuration variables is describe in Jaeger web pages.
By default a no-op tracer is created.


## Unit testing

 GO111MODULE=on go mod download
 go test ./pkg/tracelibgo

## License

See [LICENSES.txt](LICENSES.txt) file.


docs/release-notes.rst
..
.. Copyright (c) 2019 AT&T Intellectual Property.
..
.. Copyright (c) 2019 Nokia.
..
..
.. Licensed under the Creative Commons Attribution 4.0 International
..
.. Public License (the "License"); you may not use this file except
..
.. in compliance with the License. You may obtain a copy of the License at
..
..
..     https://creativecommons.org/licenses/by/4.0/
..
..
.. Unless required by applicable law or agreed to in writing, documentation
..
.. distributed under the License is distributed on an "AS IS" BASIS,
..
.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
..
.. See the License for the specific language governing permissions and
..
.. limitations under the License.
..


Release-Notes
=============


This document provides the release notes for v0.0.3 of tracelibgo.

.. contents::
   :depth: 3
   :local:


Version history
---------------

+--------------------+--------------------+--------------------+--------------------+
| **Date**           | **Ver.**           | **Author**         | **Comment**        |
|                    |                    |                    |                    |
+--------------------+--------------------+--------------------+--------------------+
| 2019-09-04         | v0.0.3             | R Riska            | First version      |
|                    |                    |                    |                    |
+--------------------+--------------------+--------------------+--------------------+


Summary
-------

This is the first version of the package.
It implements a function for creating a configured tracer instance.




Release Data
------------

This is the first version of this library.



Feature Additions
^^^^^^^^^^^^^^^^^

Bug Corrections
^^^^^^^^^^^^^^^


Deliverables
^^^^^^^^^^^^

Software Deliverables
+++++++++++++++++++++

Golang package: 
gerrit.o-ran-sc.org/ric-plt/tracelibgo/pkg/tracelibgo




Documentation Deliverables
++++++++++++++++++++++++++





Known Limitations, Issues and Workarounds
-----------------------------------------

System Limitations
^^^^^^^^^^^^^^^^^^



Known Issues
^^^^^^^^^^^^

Workarounds
^^^^^^^^^^^





References
----------




pkg/tracelibgo/tracing_test.go
/*
 * Copyright (c) 2019 AT&T Intellectual Property.
 * Copyright (c) 2018-2019 Nokia.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tracelibgo

import (
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
)

type ConfMakerTestSuite struct {
	suite.Suite
	cm confMaker
}

func (suite *ConfMakerTestSuite) SetupTest() {
	suite.cm = confMaker{"foo"}
}

func (suite *ConfMakerTestSuite) TearDownTest() {
	os.Unsetenv(tracingEnabledEnv)
	os.Unsetenv(jaegerSamplerTypeEnv)
	os.Unsetenv(jaegerSamplerParamEnv)
	os.Unsetenv(jaegerAgentAddrEnv)
	os.Unsetenv(jaegerLogLevelEnv)
}

func (suite *ConfMakerTestSuite) TestTracingIsDisabledByDefault() {
	suite.False(suite.cm.IsTracingEnabled())
}

func (suite *ConfMakerTestSuite) TestTracingCanBeEnabledWithEnvVar() {
	os.Setenv(tracingEnabledEnv, "1")
	suite.True(suite.cm.IsTracingEnabled())
	os.Setenv(tracingEnabledEnv, "true")
	suite.True(suite.cm.IsTracingEnabled())
}

func (suite *ConfMakerTestSuite) TestTracingEnabledWithUnknownValueResultsDisabled() {
	os.Setenv(tracingEnabledEnv, "0")
	suite.False(suite.cm.IsTracingEnabled())
	os.Setenv(tracingEnabledEnv, "foo")
	suite.False(suite.cm.IsTracingEnabled())
}

func (suite *ConfMakerTestSuite) TestSamplerTypeDefaultIsConst() {
	suite.Equal("const", suite.cm.getSamplerConfig().Type)
}

func (suite *ConfMakerTestSuite) TestSamplerTypeParamDefault() {
	suite.Equal(0.001, suite.cm.getSamplerConfig().Param)
}

func (suite *ConfMakerTestSuite) TestSamplerTypeCanBeDefined() {
	os.Setenv(jaegerSamplerTypeEnv, "probabilistic")
	suite.Equal("probabilistic", suite.cm.getSamplerConfig().Type)
}

func (suite *ConfMakerTestSuite) TestAgentAddrCanBeDefined() {
	os.Setenv(jaegerAgentAddrEnv, "1.1.1.1:1111")
	suite.Equal("1.1.1.1:1111", suite.cm.getReporterConfig().LocalAgentHostPort)
}

func (suite *ConfMakerTestSuite) TestAgentAddressPortIsOptional() {
	os.Setenv(jaegerAgentAddrEnv, "1.1.1.1")
	suite.Equal("1.1.1.1:6831", suite.cm.getReporterConfig().LocalAgentHostPort)
}

func (suite *ConfMakerTestSuite) TestLoggingLevelDefaultIsErr() {
	suite.Equal(logErr, suite.cm.getLoggingLevel())
}

func (suite *ConfMakerTestSuite) TestLoggingLevelCanBeConfigured() {
	os.Setenv(jaegerLogLevelEnv, "error")
	suite.Equal(logErr, suite.cm.getLoggingLevel())
	os.Setenv(jaegerLogLevelEnv, "all")
	suite.Equal(logAll, suite.cm.getLoggingLevel())
	os.Setenv(jaegerLogLevelEnv, "none")
	suite.Equal(logNone, suite.cm.getLoggingLevel())
}

func (suite *ConfMakerTestSuite) TestConfiguredTracerCreate() {
	os.Setenv(tracingEnabledEnv, "1")
	os.Setenv(jaegerSamplerParamEnv, "const")
	os.Setenv(jaegerSamplerParamEnv, "1")
	os.Setenv(jaegerAgentAddrEnv, "127.0.0.1:6831")
	tracer, closer := CreateTracer("foo")
	suite.NotNil(tracer)
	suite.NotNil(closer)
}

func (suite *ConfMakerTestSuite) TestIfTracerCreationFailsDisabledTracerIsReturned() {
	os.Setenv(tracingEnabledEnv, "1")
	tracer, closer := CreateTracer("") // Empty name is an error
	suite.NotNil(tracer)
	suite.NotNil(closer)
}

func TestConfMakerSuite(t *testing.T) {
	suite.Run(t, new(ConfMakerTestSuite))
}

func TestDefaultTracerCreate(t *testing.T) {
	tracer, closer := CreateTracer("foo")
	assert.NotNil(t, tracer)
	assert.NotNil(t, closer)
	err := closer.Close()
	assert.Nil(t, err)
}


examples/utils.go
package main

/*
#include <time.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <rmr/rmr.h>

void write_bytes_array(unsigned char *dst, void *data, int len) {
    memcpy((void *)dst, (void *)data, len);
}

#cgo CFLAGS: -I../
#cgo LDFLAGS: -lrmr_nng -lnng
*/
import "C"

import (
	"time"
	"fmt"
	"unsafe"
	"github.com/opentracing/opentracing-go"
	"encoding/json"
)

func initRmr(protoport string) unsafe.Pointer {
	protport := C.CString(protoport)
	context := C.rmr_init(protport, C.int(1000), C.int(0))
	if context == nil {
		panic("Cannot init rmr")
	}
	for {
		if ready := int(C.rmr_ready(context)); ready == 1 {
			break
		}
		time.Sleep(10*time.Second)
		fmt.Println("Wait for rmr")
	}
	return context
}

func sendMsg(rmrcontext unsafe.Pointer, mtype int, traceContext []byte) {
	var buf *C.rmr_mbuf_t

	if len(traceContext) > 0 {
		buf = C.rmr_tralloc_msg(rmrcontext, C.int(10), C.int(len(traceContext)), (*C.uchar)(unsafe.Pointer(&traceContext[0])))
	} else {
		buf = C.rmr_alloc_msg(rmrcontext, C.int(10))
	}
	buf.mtype = C.int(mtype)
	buf.sub_id = C.int(0)
	buf.len = C.int(10)
	txBuffer := C.rmr_send_msg(rmrcontext, buf)
	if txBuffer.state != C.RMR_OK {
		fmt.Println("RMR send error")
	}
}

func recvMsg(rmrcontext unsafe.Pointer) *C.rmr_mbuf_t {
	return  C.rmr_rcv_msg(rmrcontext, nil)
}

func getTraceContext(rmrbuf *C.rmr_mbuf_t) []byte{
	var traceLen C.int
	tracecont := C.rmr_trace_ref(rmrbuf, &traceLen)
	if tracecont != nil && int(traceLen) > 0 {
		return C.GoBytes(tracecont, traceLen)
	}
	return []byte{}
}

func newSpan(spanName string, traceContext []byte) opentracing.Span {
	if len(traceContext) > 0 {
		var carrier map[string]string
		err := json.Unmarshal(traceContext, &carrier)
		if err != nil {
			fmt.Println("Json unmarshal error: ", err.Error())
		} else {
			cont, err := opentracing.GlobalTracer().Extract(opentracing.TextMap,
					opentracing.TextMapCarrier(carrier))
			if err == nil {
				return opentracing.StartSpan(spanName, opentracing.ChildOf(cont))
			} else {
				fmt.Println("span extract error: ", err.Error())
			}
		}
	}
	return opentracing.StartSpan(spanName)
}

func startSpan(spanName string, traceContext []byte) (opentracing.Span, []byte) {
	span := newSpan(spanName, traceContext)
	carrier := make(map[string]string)
	opentracing.GlobalTracer().Inject(span.Context(),
				opentracing.TextMap,
			opentracing.TextMapCarrier(carrier))
	fmt.Println(carrier)
	newContext, _ := json.Marshal(carrier)
	return span, newContext
}

func freeBuffer(buf *C.rmr_mbuf_t) {
	C.rmr_free_msg(buf)
}

examples/client.go
package main

import (
	"tracelibgo/pkg/tracelibgo"
	"github.com/opentracing/opentracing-go"
	"github.com/opentracing/opentracing-go/log"
	"fmt"
	"os"
	"strconv"
	"unsafe"
)

func spanner(rmrcontext unsafe.Pointer, parentcontext []byte, counter int) {
	span, spancontext := startSpan("child :" + strconv.Itoa(counter), parentcontext)
	span.LogFields(log.Int("Counter value", counter))
	sendMsg(rmrcontext, 1, spancontext)
	buf := recvMsg(rmrcontext)
	span.Finish()
	spancontext = getTraceContext(buf)
	freeBuffer(buf)
	if (counter > 0) {
		spanner(rmrcontext, spancontext, counter-1)
	}
}


func main() {
	fmt.Println("Creating tracer")
	tracer, closer := tracelibgo.CreateTracer("Client process")
	opentracing.SetGlobalTracer(tracer)
	defer closer.Close()
	rmrcontext := initRmr("tcp:5000")
	var count int
	if len(os.Args) > 1 {
		count, _ = strconv.Atoi(os.Args[1])
	}
	if count == 0 {
		count = 1
	}
	parent, parentcontext := startSpan("Span start", []byte{})
	defer parent.Finish()
	spanner(rmrcontext, parentcontext, count)
}


pkg/tracelibgo/tracing.go
/*
 * Copyright (c) 2019 AT&T Intellectual Property.
 * Copyright (c) 2018-2019 Nokia.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Package tracelibgo implements a function to create a configured tracer instance
package tracelibgo

import (
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"

	"github.com/opentracing/opentracing-go"
	jaegercfg "github.com/uber/jaeger-client-go/config"
	jaegerlog "github.com/uber/jaeger-client-go/log"
)

const tracingEnabledEnv string = "TRACING_ENABLED"
const jaegerSamplerTypeEnv string = "TRACING_JAEGER_SAMPLER_TYPE"
const jaegerSamplerParamEnv = "TRACING_JAEGER_SAMPLER_PARAM"
const jaegerAgentAddrEnv = "TRACING_JAEGER_AGENT_ADDR"
const jaegerLogLevelEnv = "TRACING_JAEGER_LOG_LEVEL"

type confMaker struct {
	ServiceName string
}

type logLevel int

const (
	logAll  logLevel = iota
	logErr  logLevel = iota
	logNone logLevel = iota
)

func (cm *confMaker) GetEnv(envName string, defval string) (retval string) {
	retval = os.Getenv(envName)
	if retval == "" {
		retval = defval
	}
	return
}

func (cm *confMaker) IsTracingEnabled() bool {
	val := cm.GetEnv(tracingEnabledEnv, "false")
	if val == "1" || strings.ToLower(val) == "true" {
		return true
	}
	return false
}

func createDisabledTracer(name string) (opentracing.Tracer, io.Closer) {
	if name == "" {
		name = "dummy"
	}
	cfg := jaegercfg.Configuration{
		ServiceName: name,
		Disabled:    true,
	}
	tracer, closer, err := cfg.NewTracer()
	if err != nil {
		fmt.Fprintln(os.Stderr, "tracelibgo: trace creation error: ", err.Error())
	}
	return tracer, closer
}

func (cm *confMaker) getSamplerConfig() jaegercfg.SamplerConfig {
	samplerType := cm.GetEnv(jaegerSamplerTypeEnv, "const")
	param, err := strconv.ParseFloat(cm.GetEnv(jaegerSamplerParamEnv, "0.001"), 64)
	if err != nil {
		param = 0
	}
	return jaegercfg.SamplerConfig{Type: samplerType, Param: param}
}

func (cm *confMaker) getReporterConfig() jaegercfg.ReporterConfig {
	agentHostPort := cm.GetEnv(jaegerAgentAddrEnv, "127.0.0.1:6831")
	if !strings.Contains(agentHostPort, ":") {
		agentHostPort += ":6831"
	}
	return jaegercfg.ReporterConfig{LogSpans: cm.getLoggingLevel() == logAll, LocalAgentHostPort: agentHostPort}
}

func (cm *confMaker) getLoggingLevel() logLevel {
	level := strings.ToLower(cm.GetEnv(jaegerLogLevelEnv, "error"))
	switch level {
	case "error":
		return logErr
	case "all":
		return logAll
	default:
		return logNone
	}
}

// CreateTracer creates a tracer entry
func CreateTracer(name string) (opentracing.Tracer, io.Closer) {

	cm := confMaker{name}
	if !cm.IsTracingEnabled() {
		return createDisabledTracer(name)
	}
	sampler := cm.getSamplerConfig()
	reporter := cm.getReporterConfig()
	cfg := jaegercfg.Configuration{
		ServiceName: name,
		Disabled:    false,
		Sampler:     &sampler,
		Reporter:    &reporter,
	}
	var jaegerLoggerOpt jaegercfg.Option
	switch cm.getLoggingLevel() {
	case logAll, logErr:
		jaegerLoggerOpt = jaegercfg.Logger(jaegerlog.StdLogger)
	default:
		jaegerLoggerOpt = jaegercfg.Logger(nil)
	}
	tracer, closer, err := cfg.NewTracer(jaegerLoggerOpt)
	if err != nil {
		fmt.Fprintln(os.Stderr, "tracelibgo: cannot init tracer: "+err.Error())
		return createDisabledTracer(name)
	}
	return tracer, closer
}
